### ISSUE TITLE
Understanding VoxelBuffer::copy_from fail condition

### ISSUE THREAD

**Issue by afonsolage** [Zylann/godot_voxel#51]
Hello folks!

Given the method `VoxelBuffer::copy_from` bellow:
https://github.com/Zylann/godot_voxel/blob/7ced53ba6bcefefb42efe5f458d357932ccdf84a/voxel_buffer.cpp#L194-L211

Why it must ensure the previous buffer isn't equals to current one?
https://github.com/Zylann/godot_voxel/blob/7ced53ba6bcefefb42efe5f458d357932ccdf84a/voxel_buffer.cpp#L196

Reading the method, doesn't make sense it fails if the new buffer has the same size the current one. Shouldn't it be?
```cpp
 ERR_FAIL_COND(other._size != _size);
```

Thanks!

**Comment by Zylann** [comment]
Oh. Yeah that condition should be the opposite. Looks like I never used this function in the end xD

**Comment by afonsolage** [comment]
There is a similar problem (I think so) on other `VoxelBuffer::copy_from`. Given:
https://github.com/Zylann/godot_voxel/blob/7ced53ba6bcefefb42efe5f458d357932ccdf84a/voxel_buffer.cpp#L213-L260

This line:
https://github.com/Zylann/godot_voxel/blob/7ced53ba6bcefefb42efe5f458d357932ccdf84a/voxel_buffer.cpp#L229

Should be `if (other._size == _size) {` because we can only the other `VoxelBuffer::copy_from` only if the other and current one size's are equals. Even if the `area_size` is equals to the destiny of the copy, we won't provide any offset, so it'll always fail.

**Comment by Zylann** [comment]
Actually `copy_from` without arguments can only be used if `other._size == _size` AND if the area specified covers the whole block anyways.

**Comment by afonsolage** [comment]
But what if I want to copy an area that is the same size of target buffer? It will evaluate to `true` and will crash on another `copy_from`, since there is the following check: `ERR_FAIL_COND(other._size == _size);`

Let me give an use case. Let's say I have a padded voxel buffer of size (18, 18, 18). What if I want to copy only the "original-nonpadded-area" of the buffer into another buffer of size(16, 16, 16)? 

I can just call the function `copy_from` as follows:
```cpp
shoter_buffer->copy_from(**padded_buffer, Vector3i(buffer_padding), padded_buffer->get_size() - buffer_padding, Vector3i(), channel);
```

But it'll crash, because since the `area_size == _size` it'll try to make a direct copy using the other `copy_from`. But if we change the check to `other._size == _size` it'll still do the area copy and will work fine.

**Comment by Zylann** [comment]
Correcting your example, should be 2*padding:
```cpp
padded_buffer->get_size() - 2 * buffer_padding
```

Otherwise yeah what I meant is that if the two buffers have the same size AND the desired area is also that same size, then the `copy_from` without area could be used.

**Comment by afonsolage** [comment]
So the correct way should be:

```cpp
 if (area_size == _size && _other.size == _size) { 
```

Is that so?

Also, thanks for pointing me out that error, you fixed a future bug on my voxel lighting code :smile:

**Comment by Zylann** [comment]
Yes that check would be correct.
Good to hear :D

**Comment by Zylann** [comment]
Fixed in 3074f8228de35b38d438d354f4e24242936e338f
